### PR TITLE
Add type hints to test.test_retry

### PR DESCRIPTION
**Pull request by hramezani**

### CODE REVIEW

File: src/urllib3/exceptions.py [hunk 1]
```diff
@@ -32,9 +32,9 @@ class HTTPWarning(Warning):
 class PoolError(HTTPError):
     """Base exception for errors caused within a pool."""
 
-    pool: "ConnectionPool"
+    pool: Optional["ConnectionPool"]
```
Comment (bluetech) [inline src/urllib3/exceptions.py:35]: Preexisting: this line looks redundant to me, unless I'm missing something.

File: src/urllib3/exceptions.py [hunk 2]
```diff
@@ -32,9 +32,9 @@ class HTTPWarning(Warning):
...
 
-    def __init__(self, pool: "ConnectionPool", message: str) -> None:
+    def __init__(self, pool: Optional["ConnectionPool"], message: str) -> None:
```
Comment (bluetech) [inline src/urllib3/exceptions.py:37]: Given that we only need to change this due to the tests, I infer that the actual code never passes `None` here, so it's better to fix the tests instead of pessimizing the type for all users.

Maybe adding 

```py
DUMMY_POOL = ConnectionPool("dummy")
```

to the test file and passing it instead of `None`?

**Review by bluetech — COMMENTED** [review]
LGTM, except one point.

**Review by sethmlarson — COMMENTED** [review]
LGTM, I agree with @bluetech on the `pool` parameter.

File: src/urllib3/exceptions.py (hunk 1 above)
Comment (hramezani) [reply src/urllib3/exceptions.py]: hmmm, @sethmlarson Could you please elaborate on it?

File: src/urllib3/exceptions.py (hunk 2 above)
Comment (hramezani) [reply src/urllib3/exceptions.py]: Great suggestion. Thanks!

Comment (sethmlarson) [reply src/urllib3/exceptions.py]: This was the suggestion I wanted implemented :)

**Review by sethmlarson — APPROVED** [review]
